<h3>Import Libraries

<H3>IBM DATA SCIENCE PROFESSIONAL CERTIFICATE
            <h4> CAPSTONE PROJECT <h2>Opening an Indian Restaurant in Calgary
    

<h3> Introduction
    

Calgary, is one of the Canadian cosmopolitan cities famous for its status as the centre of Canada’s oil industry. It has consistently been recognized for its high quality of life. Calgary is a home for opportunities. People migrate to these beautiful city to establish new career and businesses  for a better lifestyle.

<h4> Business Problem

The aim of this capstone project is to analyze the data of various neighborhoods in Calgary and determine best locations to set up a new business "Indian Restaurants". 

<h4>Target audience for this project

Calgary is one of the fastest growing cities in Canada, with a population over 1.3 million. It is constantly developing new neighborhoods and inviting the investors to contribute to the growth of the city. This project would assist the investors and stakeholders to determine their industry for future investments.

<h4> Data required
<h5> Data needed to provide the targeted insights:
<h5> * List of neighborhoods in Calgary including new communities.
<h5> * Geographical data of these neighborhoods, like latitude and longitude information to generate map for visualization.
<h5> * Data of various businesses(venues) to apply machine learning methodology to produce insights to determine which businesses would be profitable to launch. 
    

<h4>Sources of Data
<h5> 1. Calgary Wikipedia : https://en.wikipedia.org/wiki/Calgary                            
This data is used in listing the existing neighborhoods.
<h5> 2. Open data- City of Calgary : https://data.calgary.ca/browse?tags=open+data           
This site is used in obtaining new neighborhoods and their communities.
<h5> 3. Foursquare API : https://foursquare.com/                                             
Foursquare API is used to obtain latitude, longitudes and venue data for the different neighborhoods   

In [1]:
pip install lxml

     |████████████████████████████████| 5.8MB 25.7MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [1]:
import random # library for random number generation
import numpy as np # library for vectorized computation
import pandas as pd # library to process data as dataframes
import requests # Requests allows you to send organic, grass-fed HTTP/1.1 requests
import matplotlib.pyplot as plt # plotting library
# backend for rendering plots within the browser
%matplotlib inline 
 
from sklearn.cluster import KMeans 
from sklearn.datasets.samples_generator import make_blobs

print('Libraries imported.')

Libraries imported.


<h3>Load the data of city of Calgary into pandas Dataframe

In [2]:
calgary='https://data.calgary.ca/api/views/j9ps-fyst/rows.csv'
my_table=pd.read_csv(calgary)
my_table.head(3)

CLASS  CLASS_CODE COMM_CODE      NAME     SECTOR         SRG  \
0        Residential           1       WAL    WALDEN      SOUTH  DEVELOPING   
1  Residual Sub Area           4       12C       12C  SOUTHEAST         NaN   
2        Residential           1       HSN  HASKAYNE  NORTHWEST  DEVELOPING   

  COMM_STRUCTURE   longitude   latitude                              location  
0   BUILDING OUT -114.032588  50.870404  (50.870404086764, -114.032588132238)  
1          OTHER -113.903433  50.925093  (50.925093242931, -113.903433176841)  
2   BUILDING OUT -114.286632  51.116943  (51.116943479851, -114.286631687736)

In [4]:
df=my_table[['NAME','SECTOR','SRG','longitude','latitude','location']]
df

NAME     SECTOR         SRG   longitude   latitude  \
0                    WALDEN      SOUTH  DEVELOPING -114.032588  50.870404   
1                       12C  SOUTHEAST         NaN -113.903433  50.925093   
2                  HASKAYNE  NORTHWEST  DEVELOPING -114.286632  51.116943   
3                       05G  NORTHEAST         NaN -113.916500  51.136193   
4                       13M      SOUTH         NaN -114.042754  50.856239   
..                      ...        ...         ...         ...        ...   
301               FOOTHILLS       EAST         NaN -113.981982  50.994166   
302                     13A      SOUTH         NaN -114.198193  50.914137   
303                     06A       WEST         NaN -114.229539  51.058873   
304  SCARBORO/ SUNALTA WEST     CENTRE   BUILT-OUT -114.114864  51.042196   
305           MONTEREY PARK  NORTHEAST   BUILT-OUT -113.926871  51.081460   

                                 location  
0    (50.870404086764, -114.032588132238)  
1    (50.925093242931, -113.903433176841)  
2    (51.116943479851, -114.286631687736)  
3    (51.136192848558, -113.916500322111)  
4     (50.85623929309, -114.042753841451)  
..                                    ...  
301   (50.99416609012, -113.981981981634)  
302  (50.914136620917, -114.198192846328)  
303  (51.058873158392, -114.229538699879)  
304  (51.042195714836, -114.114863752224)  
305  (51.081459900409, -113.926871436595)  

[306 rows x 6 columns]

In [5]:
df=df.dropna()
df

NAME     SECTOR         SRG   longitude   latitude  \
0                    WALDEN      SOUTH  DEVELOPING -114.032588  50.870404   
2                  HASKAYNE  NORTHWEST  DEVELOPING -114.286632  51.116943   
5                   BELMONT      SOUTH  DEVELOPING -114.055280  50.868746   
6                     SETON  SOUTHEAST  DEVELOPING -113.947448  50.873700   
8                 HOMESTEAD  NORTHEAST  DEVELOPING -113.916498  51.121630   
..                      ...        ...         ...         ...        ...   
297               CITYSCAPE  NORTHEAST  DEVELOPING -113.958381  51.147030   
298         BONAVISTA DOWNS      SOUTH   BUILT-OUT -114.031398  50.943701   
299                OAKRIDGE      SOUTH   BUILT-OUT -114.129609  50.969739   
304  SCARBORO/ SUNALTA WEST     CENTRE   BUILT-OUT -114.114864  51.042196   
305           MONTEREY PARK  NORTHEAST   BUILT-OUT -113.926871  51.081460   

                                 location  
0    (50.870404086764, -114.032588132238)  
2    (51.116943479851, -114.286631687736)  
5    (50.868746163066, -114.055280261292)  
6    (50.873700102522, -113.947448012477)  
8    (51.121629796862, -113.916498482126)  
..                                    ...  
297  (51.147029815093, -113.958380685236)  
298  (50.943700716542, -114.031397620759)  
299   (50.969739163902, -114.12960912993)  
304  (51.042195714836, -114.114863752224)  
305  (51.081459900409, -113.926871436595)  

[213 rows x 6 columns]

In [6]:
df1=df[['NAME','SECTOR','longitude','latitude','location']]
df1

NAME     SECTOR   longitude   latitude  \
0                    WALDEN      SOUTH -114.032588  50.870404   
2                  HASKAYNE  NORTHWEST -114.286632  51.116943   
5                   BELMONT      SOUTH -114.055280  50.868746   
6                     SETON  SOUTHEAST -113.947448  50.873700   
8                 HOMESTEAD  NORTHEAST -113.916498  51.121630   
..                      ...        ...         ...        ...   
297               CITYSCAPE  NORTHEAST -113.958381  51.147030   
298         BONAVISTA DOWNS      SOUTH -114.031398  50.943701   
299                OAKRIDGE      SOUTH -114.129609  50.969739   
304  SCARBORO/ SUNALTA WEST     CENTRE -114.114864  51.042196   
305           MONTEREY PARK  NORTHEAST -113.926871  51.081460   

                                 location  
0    (50.870404086764, -114.032588132238)  
2    (51.116943479851, -114.286631687736)  
5    (50.868746163066, -114.055280261292)  
6    (50.873700102522, -113.947448012477)  
8    (51.121629796862, -113.916498482126)  
..                                    ...  
297  (51.147029815093, -113.958380685236)  
298  (50.943700716542, -114.031397620759)  
299   (50.969739163902, -114.12960912993)  
304  (51.042195714836, -114.114863752224)  
305  (51.081459900409, -113.926871436595)  

[213 rows x 5 columns]

In [7]:
df1.sort_values(by=['SECTOR']).reset_index(drop=True)

NAME  SECTOR   longitude   latitude  \
0    DOWNTOWN EAST VILLAGE  CENTRE -114.050817  51.047065   
1                BRITANNIA  CENTRE -114.086324  51.012670   
2                  MISSION  CENTRE -114.066322  51.031090   
3               ELBOW PARK  CENTRE -114.083844  51.021364   
4        DOWNTOWN WEST END  CENTRE -114.087788  51.047966   
..                     ...     ...         ...        ...   
208            ASPEN WOODS    WEST -114.207901  51.045129   
209          CHRISTIE PARK    WEST -114.175901  51.040843   
210              PATTERSON    WEST -114.177047  51.063838   
211          MEDICINE HILL    WEST -114.200053  51.079291   
212        DISCOVERY RIDGE    WEST -114.216583  51.015482   

                                 location  
0    (51.047064817249, -114.050817003339)  
1    (51.012670378595, -114.086323863092)  
2      (51.0310903238, -114.066321693987)  
3    (51.021363851538, -114.083843633682)  
4     (51.047966240306, -114.08778794454)  
..                                    ...  
208  (51.045129216321, -114.207900713863)  
209  (51.040843232182, -114.175900760572)  
210  (51.063837750822, -114.177046508603)  
211  (51.079291077961, -114.200053267025)  
212  (51.015482335131, -114.216582665462)  

[213 rows x 5 columns]

In [8]:
df1=df1.rename(columns={"NAME":'Neighborhood'})
df1.head()

Neighborhood     SECTOR   longitude   latitude  \
0       WALDEN      SOUTH -114.032588  50.870404   
2     HASKAYNE  NORTHWEST -114.286632  51.116943   
5      BELMONT      SOUTH -114.055280  50.868746   
6        SETON  SOUTHEAST -113.947448  50.873700   
8    HOMESTEAD  NORTHEAST -113.916498  51.121630   

                               location  
0  (50.870404086764, -114.032588132238)  
2  (51.116943479851, -114.286631687736)  
5  (50.868746163066, -114.055280261292)  
6  (50.873700102522, -113.947448012477)  
8  (51.121629796862, -113.916498482126)

In [9]:
print('The dataframe has {} sectors and {} neighborhoods.'.format(
        len(df1['SECTOR'].unique()),
        df1.shape[0]
    )
)

The dataframe has 8 sectors and 213 neighborhoods.


<h3>Dataframe is clean and ready for futher analysis.

<h3>Import Libraries

In [10]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

print('Folium installed')
print('Libraries imported.')

Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.5.11
  latest version: 4.7.12

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs: 
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geopy-1.20.0               |             py_0          57 KB  conda-forge
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          91 KB

The following NEW packages will be INSTALLED:

    geographiclib: 1.50-py_0   conda-forge
    geopy:         1.20.0-py_0 conda-forge


geopy-1.20.0         | 57 KB     | ##################################### | 100% 
geographiclib-1.50   | 

<h3>Using Four Square API to obtain Latitude and Longitude information for the city of Calgary

In [11]:
CLIENT_ID = '4ZLNJ532NDROLNRSQIRCHHU4CQY22WQZAILNX5QF4ZB5T1SS' # your Foursquare ID
CLIENT_SECRET = 'HLES54GBQUAFEJFI4YS4MXSEJ4BAG5EZ34JQ5DC3ANTFLOPG' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 4ZLNJ532NDROLNRSQIRCHHU4CQY22WQZAILNX5QF4ZB5T1SS
CLIENT_SECRET:HLES54GBQUAFEJFI4YS4MXSEJ4BAG5EZ34JQ5DC3ANTFLOPG


In [12]:
address = 'Calgary'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(latitude, longitude)

51.02532675 -114.049868485806


<h3>Map showing the neighborhoods of Calgary

In [13]:
# create map of Calgary using latitude and longitude values
map_calgary = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map|
for latitude, longitude, Neighborhood, SECTOR in zip(df1['latitude'], df1['longitude'], df1['SECTOR'], df1['Neighborhood']):
    label = '{}, {}'.format(Neighborhood, SECTOR)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_calgary)  
    
map_calgary

<h3>Extracting venue data for all the neighborhoods of Calgary within a radius of 1000 and limit of 100 

In [17]:
radius = 1000
LIMIT = 100

venues = []

for lat, long, neighborhood in zip(df1['latitude'], df1['longitude'], df1['Neighborhood']):
    
    # create the API request URL
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        LIMIT)
    
    # make the GET request
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    # return only relevant information for each nearby venue
    for venue in results:
        venues.append((
            neighborhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [18]:
# convert the venues list into a new DataFrame
venues_df = pd.DataFrame(venues)

# define the column names
venues_df.columns = ['Neighborhood', 'Latitude', 'Longitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']

print(venues_df.shape)
venues_df.head()

(4359, 7)


Neighborhood   Latitude   Longitude                   VenueName  \
0       WALDEN  50.870404 -114.032588               Save-On-Foods   
1       WALDEN  50.870404 -114.032588                   Starbucks   
2       WALDEN  50.870404 -114.032588  Spruce It Up Garden Centre   
3       WALDEN  50.870404 -114.032588                Windsor Rose   
4       WALDEN  50.870404 -114.032588                 Tim Hortons   

   VenueLatitude  VenueLongitude  VenueCategory  
0      50.874971     -114.038745  Grocery Store  
1      50.875425     -114.036857    Coffee Shop  
2      50.864841     -114.038754  Garden Center  
3      50.875968     -114.037497      Gastropub  
4      50.875371     -114.036337    Coffee Shop

<h3> Venues for each Neighborhood

In [19]:
venues_df.groupby(["Neighborhood"]).count()

Latitude  Longitude  VenueName  VenueLatitude  \
Neighborhood                                                                  
01F                                  3          3          3              3   
ABBEYDALE                            4          4          4              4   
ACADIA                              14         14         14             14   
ALBERT PARK/RADISSON HEIGHTS        14         14         14             14   
ALTADORE                            16         16         16             16   
...                                ...        ...        ...            ...   
WINSTON HEIGHTS/MOUNTVIEW            6          6          6              6   
WOLF WILLOW                          3          3          3              3   
WOODBINE                             8          8          8              8   
WOODLANDS                            8          8          8              8   
YORKVILLE                            8          8          8              8   

                              VenueLongitude  VenueCategory  
Neighborhood                                                 
01F                                        3              3  
ABBEYDALE                                  4              4  
ACADIA                                    14             14  
ALBERT PARK/RADISSON HEIGHTS              14             14  
ALTADORE                                  16             16  
...                                      ...            ...  
WINSTON HEIGHTS/MOUNTVIEW                  6              6  
WOLF WILLOW                                3              3  
WOODBINE                                   8              8  
WOODLANDS                                  8              8  
YORKVILLE                                  8              8  

[207 rows x 6 columns]

<h3> Unique categories of all the returned values

In [22]:
print('There are {} uniques categories.'.format(len(venues_df['VenueCategory'].unique())))

There are 260 uniques categories.


In [23]:
# print out the list of categories
venues_df['VenueCategory'].unique()[:50]


array(['Grocery Store', 'Coffee Shop', 'Garden Center', 'Gastropub',
       'Bank', 'Fast Food Restaurant', 'Pharmacy', 'Indian Restaurant',
       'Mexican Restaurant', 'Construction & Landscaping',
       'Movie Theater', 'Supermarket', 'Café', 'Recreation Center',
       'Liquor Store', 'Athletics & Sports', 'Hotel', 'ATM',
       'Health & Beauty Service', 'Gym', 'Playground', 'Park',
       'Ice Cream Shop', 'Food & Drink Shop', 'Brewery', 'Restaurant',
       'Vietnamese Restaurant', 'Music Store', 'Korean Restaurant', 'Pub',
       'Bagel Shop', 'Smoke Shop', 'Snack Place', 'Sushi Restaurant',
       'Bar', 'American Restaurant', 'Shopping Mall', 'Comic Shop',
       'Rental Car Location', 'Wine Shop', 'Golf Course',
       'Convenience Store', 'Pizza Place', 'Lake', 'Sandwich Place',
       'Chinese Restaurant', 'Skating Rink', 'Video Store', 'Theme Park',
       'Electronics Store'], dtype=object)

In [24]:
# check if the results contain "Indian Restaurant"
"Neighborhood" in venues_df['VenueCategory'].unique()

True

<h3> Analyze each neighborhood

In [25]:
# one hot encoding
cal_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
cal_onehot['Neighborhoods'] = venues_df['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [cal_onehot.columns[-1]] + list(cal_onehot.columns[:-1])
cal_onehot = cal_onehot[fixed_columns]

print(cal_onehot.shape)
cal_onehot.head()



(4359, 261)


Neighborhoods  ATM  Accessories Store  American Restaurant  Arcade  \
0        WALDEN    0                  0                    0       0   
1        WALDEN    0                  0                    0       0   
2        WALDEN    0                  0                    0       0   
3        WALDEN    0                  0                    0       0   
4        WALDEN    0                  0                    0       0   

   Argentinian Restaurant  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0                       0            0                    0                 0   
1                       0            0                    0                 0   
2                       0            0                    0                 0   
3                       0            0                    0                 0   
4                       0            0                    0                 0   

   Athletics & Sports  ...  Vietnamese Restaurant  Warehouse Store  \
0                   0  ...                      0                0   
1                   0  ...                      0                0   
2                   0  ...                      0                0   
3                   0  ...                      0                0   
4                   0  ...                      0                0   

   Water Park  Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  \
0           0         0          0            0              0            0   
1           0         0          0            0              0            0   
2           0         0          0            0              0            0   
3           0         0          0            0              0            0   
4           0         0          0            0              0            0   

   Zoo  Zoo Exhibit  
0    0            0  
1    0            0  
2    0            0  
3    0            0  
4    0            0  

[5 rows x 261 columns]

<h3> Groupby Neighborhood by taking the mean of the frequency of occurances of each category

In [26]:
cal_grouped = cal_onehot.groupby(["Neighborhoods"]).mean().reset_index()

print(cal_grouped.shape)
cal_grouped

(207, 261)


Neighborhoods  ATM  Accessories Store  \
0                             01F  0.0                0.0   
1                       ABBEYDALE  0.0                0.0   
2                          ACADIA  0.0                0.0   
3    ALBERT PARK/RADISSON HEIGHTS  0.0                0.0   
4                        ALTADORE  0.0                0.0   
..                            ...  ...                ...   
202     WINSTON HEIGHTS/MOUNTVIEW  0.0                0.0   
203                   WOLF WILLOW  0.0                0.0   
204                      WOODBINE  0.0                0.0   
205                     WOODLANDS  0.0                0.0   
206                     YORKVILLE  0.0                0.0   

     American Restaurant  Arcade  Argentinian Restaurant  Art Gallery  \
0               0.000000     0.0                     0.0          0.0   
1               0.000000     0.0                     0.0          0.0   
2               0.071429     0.0                     0.0          0.0   
3               0.000000     0.0                     0.0          0.0   
4               0.000000     0.0                     0.0          0.0   
..                   ...     ...                     ...          ...   
202             0.000000     0.0                     0.0          0.0   
203             0.000000     0.0                     0.0          0.0   
204             0.000000     0.0                     0.0          0.0   
205             0.000000     0.0                     0.0          0.0   
206             0.000000     0.0                     0.0          0.0   

     Arts & Crafts Store  Asian Restaurant  Athletics & Sports  ...  \
0               0.000000          0.000000                 0.0  ...   
1               0.000000          0.000000                 0.0  ...   
2               0.071429          0.000000                 0.0  ...   
3               0.000000          0.071429                 0.0  ...   
4               0.000000          0.000000                 0.0  ...   
..                   ...               ...                 ...  ...   
202             0.000000          0.000000                 0.0  ...   
203             0.000000          0.000000                 0.0  ...   
204             0.000000          0.125000                 0.0  ...   
205             0.000000          0.000000                 0.0  ...   
206             0.000000          0.125000                 0.0  ...   

     Vietnamese Restaurant  Warehouse Store  Water Park  Wine Bar  Wine Shop  \
0                 0.000000              0.0         0.0       0.0        0.0   
1                 0.000000              0.0         0.0       0.0        0.0   
2                 0.000000              0.0         0.0       0.0        0.0   
3                 0.071429              0.0         0.0       0.0        0.0   
4                 0.000000              0.0         0.0       0.0        0.0   
..                     ...              ...         ...       ...        ...   
202               0.000000              0.0         0.0       0.0        0.0   
203               0.000000              0.0         0.0       0.0        0.0   
204               0.000000              0.0         0.0       0.0        0.0   
205               0.000000              0.0         0.0       0.0        0.0   
206               0.000000              0.0         0.0       0.0        0.0   

     Wings Joint  Women's Store  Yoga Studio  Zoo  Zoo Exhibit  
0            0.0            0.0          0.0  0.0          0.0  
1            0.0            0.0          0.0  0.0          0.0  
2            0.0            0.0          0.0  0.0          0.0  
3            0.0            0.0          0.0  0.0          0.0  
4            0.0            0.0          0.0  0.0          0.0  
..           ...            ...          ...  ...          ...  
202          0.0            0.0          0.0  0.0          0.0  
203          0.0            0.0          0.0  0.0          0.0  
204        

In [27]:
len(cal_grouped[cal_grouped["Indian Restaurant"] > 0])


31

<h3> New dataframe for Indian Restaurant data 

In [28]:
cal_ind = cal_grouped[["Neighborhoods","Indian Restaurant"]]
cal_ind.head()

Neighborhoods  Indian Restaurant
0                           01F           0.000000
1                     ABBEYDALE           0.000000
2                        ACADIA           0.000000
3  ALBERT PARK/RADISSON HEIGHTS           0.071429
4                      ALTADORE           0.000000

<h3>Methodology
    <h4>Clustering the neighborhoods in Calgary
    <h5>Run k-means to cluster the neighborhoods in Calgary into 5 clusters.



In [29]:
# set number of clusters
kclusters = 5

cal_clustering = cal_ind.drop(["Neighborhoods"], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(cal_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 2, 1, 1, 3, 1, 1, 1], dtype=int32)

In [30]:
# create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.
cal_merged = cal_ind.copy()

# add clustering labels
cal_merged["Cluster Labels"] = kmeans.labels_

In [31]:
cal_merged.rename(columns={"Neighborhoods": "Neighborhood"}, inplace=True)
cal_merged.head()

Neighborhood  Indian Restaurant  Cluster Labels
0                           01F           0.000000               1
1                     ABBEYDALE           0.000000               1
2                        ACADIA           0.000000               1
3  ALBERT PARK/RADISSON HEIGHTS           0.071429               2
4                      ALTADORE           0.000000               1

In [33]:
# merge cal_grouped with df1 to add latitude/longitude for each neighborhood
cal_merged = cal_merged.join(df1.set_index("Neighborhood"), on="Neighborhood")

print(cal_merged.shape)
cal_merged.head()

(207, 7)


Neighborhood  Indian Restaurant  Cluster Labels     SECTOR  \
0                           01F           0.000000               1  NORTHWEST   
1                     ABBEYDALE           0.000000               1  NORTHEAST   
2                        ACADIA           0.000000               1      SOUTH   
3  ALBERT PARK/RADISSON HEIGHTS           0.071429               2       EAST   
4                      ALTADORE           0.000000               1     CENTRE   

    longitude   latitude                              location  
0 -114.263366  51.119619  (51.119618566467, -114.263365862096)  
1 -113.927793  51.059517  (51.059516758039, -113.927792953149)  
2 -114.053702  50.972407   (50.972407239131, -114.05370183694)  
3 -113.996784  51.044542  (51.044541877637, -113.996783813916)  
4 -114.101859  51.015953  (51.015953319087, -114.101858876217)

In [34]:
# sort the results by Cluster Labels
print(cal_merged.shape)
cal_merged.sort_values(["Cluster Labels"], inplace=True)
cal_merged

(207, 7)


Neighborhood  Indian Restaurant  Cluster Labels     SECTOR  \
124          NORTH HAVEN           0.200000               0      NORTH   
87              HIGHWOOD           0.200000               0     CENTRE   
121         NEW BRIGHTON           0.200000               0  SOUTHEAST   
0                    01F           0.000000               1  NORTHWEST   
125    NORTH HAVEN UPPER           0.000000               1      NORTH   
..                   ...                ...             ...        ...   
186          TUXEDO PARK           0.024390               3     CENTRE   
138  QUEENS PARK VILLAGE           0.166667               4     CENTRE   
39           COPPERFIELD           0.142857               4  SOUTHEAST   
27           CASTLERIDGE           0.131579               4  NORTHEAST   
175      SPRINGBANK HILL           0.166667               4       WEST   

      longitude   latitude                              location  
124 -114.087941  51.097005  (51.097004668982, -114.087941023986)  
87  -114.079905  51.092354  (51.092354473106, -114.079905386226)  
121 -113.945246  50.922048  (50.922047552219, -113.945245965549)  
0   -114.263366  51.119619  (51.119618566467, -114.263365862096)  
125 -114.087063  51.103216  (51.103216351455, -114.087062763774)  
..          ...        ...                                   ...  
186 -114.061043  51.074245  (51.074245266517, -114.061043204102)  
138 -114.077609  51.085147   (51.085146679092, -114.07760880425)  
39  -113.929446  50.917810  (50.917810348178, -113.929446212339)  
27  -113.961030  51.105257  (51.105257372741, -113.961029784231)  
175 -114.208454  51.026835  (51.026835148901, -114.208453682797)  

[207 rows x 7 columns]

<h3>Visualize the clusters

<h3>Import Libraries

In [36]:
import folium
import matplotlib.pyplot as plt
import matplotlib.cm as cm
import matplotlib.colors as colors

In [38]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(cal_merged['latitude'], cal_merged['longitude'], cal_merged['Neighborhood'], cal_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' - Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In [39]:
# save the map as HTML file
map_clusters.save('map_clusters.html')

<h3>Examining the Clusters
    <h4>Cluster 0

In [40]:
cal_merged.loc[cal_merged['Cluster Labels'] == 0]

Neighborhood  Indian Restaurant  Cluster Labels     SECTOR   longitude  \
124   NORTH HAVEN                0.2               0      NORTH -114.087941   
87       HIGHWOOD                0.2               0     CENTRE -114.079905   
121  NEW BRIGHTON                0.2               0  SOUTHEAST -113.945246   

      latitude                              location  
124  51.097005  (51.097004668982, -114.087941023986)  
87   51.092354  (51.092354473106, -114.079905386226)  
121  50.922048  (50.922047552219, -113.945245965549)

In [52]:
cal_merged.sort_values(by=['SECTOR']).loc[cal_merged['Cluster Labels'] == 0].reset_index(drop=True)

Neighborhood  Indian Restaurant  Cluster Labels     SECTOR   longitude  \
0      HIGHWOOD                0.2               0     CENTRE -114.079905   
1   NORTH HAVEN                0.2               0      NORTH -114.087941   
2  NEW BRIGHTON                0.2               0  SOUTHEAST -113.945246   

    latitude                              location  
0  51.092354  (51.092354473106, -114.079905386226)  
1  51.097005  (51.097004668982, -114.087941023986)  
2  50.922048  (50.922047552219, -113.945245965549)

 <h4>Cluster 1

In [53]:
cal_merged.sort_values(by=['SECTOR']).loc[cal_merged['Cluster Labels'] == 1].reset_index(drop=True)

Neighborhood  Indian Restaurant  Cluster Labels  SECTOR  \
0                  ELBOYA                0.0               1  CENTRE   
1    BRIDGELAND/RIVERSIDE                0.0               1  CENTRE   
2                 MAYFAIR                0.0               1  CENTRE   
3     ST. ANDREWS HEIGHTS                0.0               1  CENTRE   
4            CAPITOL HILL                0.0               1  CENTRE   
..                    ...                ...             ...     ...   
174       CURRIE BARRACKS                0.0               1    WEST   
175       STRATHCONA PARK                0.0               1    WEST   
176              WILDWOOD                0.0               1    WEST   
177              WESTGATE                0.0               1    WEST   
178             GLAMORGAN                0.0               1    WEST   

      longitude   latitude                              location  
0   -114.073623  51.012993  (51.012992811701, -114.073623214314)  
1   -114.035074  51.050566  (51.050566015095, -114.035074421661)  
2   -114.088959  50.995804  (50.995804085063, -114.088959076983)  
3   -114.127418  51.063906    (51.06390635975, -114.12741807028)  
4   -114.094650  51.072666  (51.072666245106, -114.094650000997)  
..          ...        ...                                   ...  
174 -114.124763  51.015844   (51.01584422593, -114.124762705196)  
175 -114.177555  51.047388   (51.04738835463, -114.177555126009)  
176 -114.158522  51.058251  (51.058251224809, -114.158521596925)  
177 -114.158457  51.044160   (51.044160236555, -114.15845717807)  
178 -114.152863  51.015513   (51.01551339158, -114.152862528799)  

[179 rows x 7 columns]

 <h4>Cluster 2

In [54]:
cal_merged.sort_values(by=['SECTOR']).loc[cal_merged['Cluster Labels'] == 2].reset_index(drop=True)

Neighborhood  Indian Restaurant  Cluster Labels     SECTOR  \
0                      SCARBORO           0.076923               2     CENTRE   
1                      BANKVIEW           0.062500               2     CENTRE   
2              CAMBRIAN HEIGHTS           0.090909               2     CENTRE   
3                     SOUTHVIEW           0.071429               2       EAST   
4  ALBERT PARK/RADISSON HEIGHTS           0.071429               2       EAST   
5                PANORAMA HILLS           0.083333               2      NORTH   
6                   FALCONRIDGE           0.058824               2  NORTHEAST   
7                        RUNDLE           0.076923               2  NORTHEAST   
8                        WALDEN           0.090909               2      SOUTH   

    longitude   latitude                              location  
0 -114.107154  51.040725  (51.040724610494, -114.107153533341)  
1 -114.100947  51.034133   (51.034132750948, -114.10094728858)  
2 -114.087099  51.084594   (51.084593832377, -114.08709920912)  
3 -113.997338  51.034156  (51.034156041566, -113.997337917561)  
4 -113.996784  51.044542  (51.044541877637, -113.996783813916)  
5 -114.088066  51.159713   (51.159712816162, -114.08806596268)  
6 -113.946709  51.103299  (51.103299376072, -113.946709148585)  
7 -113.970058  51.074241  (51.074240559387, -113.970058115068)  
8 -114.032588  50.870404  (50.870404086764, -114.032588132238)

 <h4>Cluster 3

In [55]:
cal_merged.sort_values(by=['SECTOR']).loc[cal_merged['Cluster Labels'] == 3].reset_index(drop=True)

Neighborhood  Indian Restaurant  Cluster Labels     SECTOR  \
0         TUXEDO PARK           0.024390               3     CENTRE   
1           SUNNYSIDE           0.021505               3     CENTRE   
2             ROXBORO           0.024390               3     CENTRE   
3             SUNALTA           0.044444               3     CENTRE   
4             MISSION           0.017544               3     CENTRE   
5              ERLTON           0.022727               3     CENTRE   
6      CLIFF BUNGALOW           0.020000               3     CENTRE   
7   LOWER MOUNT ROYAL           0.020000               3     CENTRE   
8         RIDEAU PARK           0.034483               3     CENTRE   
9       COUNTRY HILLS           0.045455               3      NORTH   
10            KINCORA           0.037037               3      NORTH   
11        ARBOUR LAKE           0.022727               3  NORTHWEST   

     longitude   latitude                              location  
0  -114.061043  51.074245  (51.074245266517, -114.061043204102)  
1  -114.079017  51.056419  (51.056419411247, -114.079016652198)  
2  -114.067630  51.026345  (51.026344907744, -114.067629621076)  
3  -114.101455  51.042956  (51.042956146485, -114.101455470711)  
4  -114.066322  51.031090    (51.0310903238, -114.066321693987)  
5  -114.063143  51.026329    (51.02632870102, -114.06314254064)  
6  -114.074891  51.033930  (51.033929921661, -114.074890819558)  
7  -114.086500  51.036483   (51.036483393667, -114.08650004594)  
8  -114.074158  51.024998  (51.024998484418, -114.074158498928)  
9  -114.079885  51.143769  (51.143768799819, -114.079884606984)  
10 -114.133971  51.160285  (51.160285038674, -114.133970781784)  
11 -114.207593  51.132624  (51.132624336028, -114.207593370651)

 <h4>Cluster 4

In [56]:
cal_merged.sort_values(by=['SECTOR']).loc[cal_merged['Cluster Labels'] == 4].reset_index(drop=True)

Neighborhood  Indian Restaurant  Cluster Labels     SECTOR  \
0  QUEENS PARK VILLAGE           0.166667               4     CENTRE   
1          CASTLERIDGE           0.131579               4  NORTHEAST   
2          COPPERFIELD           0.142857               4  SOUTHEAST   
3      SPRINGBANK HILL           0.166667               4       WEST   

    longitude   latitude                              location  
0 -114.077609  51.085147   (51.085146679092, -114.07760880425)  
1 -113.961030  51.105257  (51.105257372741, -113.961029784231)  
2 -113.929446  50.917810  (50.917810348178, -113.929446212339)  
3 -114.208454  51.026835  (51.026835148901, -114.208453682797)

<h3>Observations
<h4>Cluster 0 : This cluster 3 neighborhoods and has equal number of Indian restaurants.
<h4>Cluster 1 : It can be observed that about 179 neighborhoods in Calgary dont have a signification number of Indian restaurants.
It is clear that starting a new Indian restaurant in these neighborhoods is a good options for new investors.    
<h4>Cluster 2 : This cluster 9 neighborhoods and has Indian restaurants.
<h4>Cluster 3 : This cluster 12 neighborhoods and has Indian restaurants.
<h4>Cluster 4 : This cluster 4 neighborhoods and Indian restaurants are present in these neighbourhoods.

In [61]:
cluster1=cal_merged.sort_values(by=['SECTOR']).loc[cal_merged['Cluster Labels'] == 1].reset_index(drop=True)
cluster1
clust1=cluster1[['Neighborhood','Indian Restaurant','SECTOR']]
clust1

Neighborhood  Indian Restaurant  SECTOR
0                  ELBOYA                0.0  CENTRE
1    BRIDGELAND/RIVERSIDE                0.0  CENTRE
2                 MAYFAIR                0.0  CENTRE
3     ST. ANDREWS HEIGHTS                0.0  CENTRE
4            CAPITOL HILL                0.0  CENTRE
..                    ...                ...     ...
174       CURRIE BARRACKS                0.0    WEST
175       STRATHCONA PARK                0.0    WEST
176              WILDWOOD                0.0    WEST
177              WESTGATE                0.0    WEST
178             GLAMORGAN                0.0    WEST

[179 rows x 3 columns]

<h2>Results
    <h3>Neighborhoods in Calgary that can become a potential area to start new Indian Restaurants 
    <h4> Data shows that 40 neighbouhoods in Central Calgary, 25 in the West, 8 in the East, 16 in the North, 38 in the South, 9 southeast, 23 northwest and 18 northeast are open for starting new Indian restaurant.
Calgary is definitly a city with immense opportunities.

In [65]:
centre = clust1[clust1.SECTOR == 'CENTRE'].reset_index(drop=True)
centre

Neighborhood  Indian Restaurant  SECTOR
0                          ELBOYA           0.000000  CENTRE
1            BRIDGELAND/RIVERSIDE           0.000000  CENTRE
2                         MAYFAIR           0.000000  CENTRE
3             ST. ANDREWS HEIGHTS           0.000000  CENTRE
4                    CAPITOL HILL           0.000000  CENTRE
5                 MEADOWLARK PARK           0.000000  CENTRE
6                       GREENVIEW           0.000000  CENTRE
7                   SOUTH CALGARY           0.000000  CENTRE
8                      ELBOW PARK           0.000000  CENTRE
9                        ROSEDALE           0.000000  CENTRE
10                       ROSEMONT           0.000000  CENTRE
11       DOWNTOWN COMMERCIAL CORE           0.000000  CENTRE
12                     SHAGANAPPI           0.000000  CENTRE
13              UPPER MOUNT ROYAL           0.012987  CENTRE
14          DOWNTOWN EAST VILLAGE           0.010000  CENTRE
15                  HIGHLAND PARK           0.000000  CENTRE
16                      HILLHURST           0.000000  CENTRE
17         SCARBORO/ SUNALTA WEST           0.000000  CENTRE
18              DOWNTOWN WEST END           0.000000  CENTRE
19                     EAU CLAIRE           0.000000  CENTRE
20                 GARRISON WOODS           0.000000  CENTRE
21                   WINDSOR PARK           0.000000  CENTRE
22      WINSTON HEIGHTS/MOUNTVIEW           0.000000  CENTRE
23                 WEST HILLHURST           0.000000  CENTRE
24            KILLARNEY/GLENGARRY           0.000000  CENTRE
25                      INGLEWOOD           0.000000  CENTRE
26                     MANCHESTER           0.000000  CENTRE
27                      BRITANNIA           0.000000  CENTRE
28                      CHINATOWN           0.000000  CENTRE
29               CRESCENT HEIGHTS           0.013333  CENTRE
30                       ALTADORE           0.000000  CENTRE
31                       PARKDALE           0.000000  CENTRE
32                       PARKHILL           0.000000  CENTRE
33                    POINT MCKAY           0.000000  CENTRE
34                         RAMSAY           0.000000  CENTRE
35  HOUNSFIELD HEIGHTS/BRIAR HILL           0.000000  CENTRE
36                       BELTLINE           0.000000  CENTRE
37                 MOUNT PLEASANT           0.000000  CENTRE
38                       BEL-AIRE           0.000000  CENTRE
39                        RENFREW           0.000000  CENTRE
40                       RICHMOND           0.000000  CENTRE
41                    BANFF TRAIL           0.000000  CENTRE

In [66]:
west = clust1[clust1.SECTOR == 'WEST'].reset_index(drop=True)
west

Neighborhood  Indian Restaurant SECTOR
0             PATTERSON                0.0   WEST
1          LINCOLN PARK                0.0   WEST
2         MEDICINE HILL                0.0   WEST
3   NORTH GLENMORE PARK                0.0   WEST
4           ROSSCARROCK                0.0   WEST
5          WEST SPRINGS                0.0   WEST
6             CRESTMONT                0.0   WEST
7          COUGAR RIDGE                0.0   WEST
8           ASPEN WOODS                0.0   WEST
9            COACH HILL                0.0   WEST
10        CHRISTIE PARK                0.0   WEST
11             LAKEVIEW                0.0   WEST
12         SPRUCE CLIFF                0.0   WEST
13      DISCOVERY RIDGE                0.0   WEST
14         VALLEY RIDGE                0.0   WEST
15       GARRISON GREEN                0.0   WEST
16            GLENBROOK                0.0   WEST
17             GLENDALE                0.0   WEST
18          SIGNAL HILL                0.0   WEST
19         RUTLAND PARK                0.0   WEST
20      CURRIE BARRACKS                0.0   WEST
21      STRATHCONA PARK                0.0   WEST
22             WILDWOOD                0.0   WEST
23             WESTGATE                0.0   WEST
24            GLAMORGAN                0.0   WEST

In [67]:
east = clust1[clust1.SECTOR == 'EAST'].reset_index(drop=True)
east

Neighborhood  Indian Restaurant SECTOR
0          BELVEDERE                0.0   EAST
1        FOREST LAWN                0.0   EAST
2     FOREST HEIGHTS                0.0   EAST
3         RED CARPET                0.0   EAST
4         ERIN WOODS                0.0   EAST
5              DOVER                0.0   EAST
6  PENBROOKE MEADOWS                0.0   EAST
7     APPLEWOOD PARK                0.0   EAST

In [68]:
NORTH = clust1[clust1.SECTOR == 'NORTH'].reset_index(drop=True)
NORTH

Neighborhood  Indian Restaurant SECTOR
0   COUNTRY HILLS VILLAGE                0.0  NORTH
1        SANDSTONE VALLEY                0.0  NORTH
2               SAGE HILL                0.0  NORTH
3           HARVEST HILLS                0.0  NORTH
4      BEDDINGTON HEIGHTS                0.0  NORTH
5           HIDDEN VALLEY                0.0  NORTH
6          COVENTRY HILLS                0.0  NORTH
7                SHERWOOD                0.0  NORTH
8        HUNTINGTON HILLS                0.0  NORTH
9       NORTH HAVEN UPPER                0.0  NORTH
10             CARRINGTON                0.0  NORTH
11               EVANSTON                0.0  NORTH
12            THORNCLIFFE                0.0  NORTH
13           MACEWAN GLEN                0.0  NORTH
14             LIVINGSTON                0.0  NORTH
15             NOLAN HILL                0.0  NORTH

In [69]:
SOUTH = clust1[clust1.SECTOR == 'SOUTH'].reset_index(drop=True)
SOUTH

Neighborhood  Indian Restaurant SECTOR
0         SHAWNESSY                0.0  SOUTH
1          HAYSBORO                0.0  SOUTH
2    SHAWNEE SLOPES                0.0  SOUTH
3         WOODLANDS                0.0  SOUTH
4         SILVERADO                0.0  SOUTH
5       MAPLE RIDGE                0.0  SOUTH
6         CEDARBRAE                0.0  SOUTH
7          SUNDANCE                0.0  SOUTH
8          WOODBINE                0.0  SOUTH
9       WOLF WILLOW                0.0  SOUTH
10      WILLOW PARK                0.0  SOUTH
11         FAIRVIEW                0.0  SOUTH
12      EAGLE RIDGE                0.0  SOUTH
13        MIDNAPORE                0.0  SOUTH
14         MILLRISE                0.0  SOUTH
15        EVERGREEN                0.0  SOUTH
16     KELVIN GROVE                0.0  SOUTH
17        SOUTHWOOD                0.0  SOUTH
18         SOMERSET                0.0  SOUTH
19       QUEENSLAND                0.0  SOUTH
20       DEER RIDGE                0.0  SOUTH
21   CANYON MEADOWS                0.0  SOUTH
22           ACADIA                0.0  SOUTH
23        KINGSLAND                0.0  SOUTH
24          BAYVIEW                0.0  SOUTH
25         OAKRIDGE                0.0  SOUTH
26  BONAVISTA DOWNS                0.0  SOUTH
27        CHAPARRAL                0.0  SOUTH
28         BRAESIDE                0.0  SOUTH
29   LAKE BONAVISTA                0.0  SOUTH
30     CHINOOK PARK                0.0  SOUTH
31         PALLISER                0.0  SOUTH
32         PARKLAND                0.0  SOUTH
33     DIAMOND COVE                0.0  SOUTH
34        PUMP HILL                0.0  SOUTH
35         DEER RUN                0.0  SOUTH
36        YORKVILLE                0.0  SOUTH
37       BRIDLEWOOD                0.0  SOUTH

In [70]:
southeast = clust1[clust1.SECTOR == 'SOUTHEAST'].reset_index(drop=True)
southeast

Neighborhood  Indian Restaurant     SECTOR
0         RIVERBEND                0.0  SOUTHEAST
1          CRANSTON                0.0  SOUTHEAST
2  DOUGLASDALE/GLEN                0.0  SOUTHEAST
3     MCKENZIE LAKE                0.0  SOUTHEAST
4             OGDEN                0.0  SOUTHEAST
5    MCKENZIE TOWNE                0.0  SOUTHEAST
6             SETON                0.0  SOUTHEAST
7          MAHOGANY                0.0  SOUTHEAST
8        AUBURN BAY                0.0  SOUTHEAST

In [72]:
northwest = clust1[clust1.SECTOR == 'NORTHWEST'].reset_index(drop=True)
northwest

Neighborhood  Indian Restaurant     SECTOR
0                EDGEMONT                0.0  NORTHWEST
1               BRENTWOOD                0.0  NORTHWEST
2                HASKAYNE                0.0  NORTHWEST
3                HAMPTONS                0.0  NORTHWEST
4    GREENWOOD/GREENBRIAR                0.0  NORTHWEST
5                 BOWNESS                0.0  NORTHWEST
6                HAWKWOOD                0.0  NORTHWEST
7               DALHOUSIE                0.0  NORTHWEST
8                 CITADEL                0.0  NORTHWEST
9             CHARLESWOOD                0.0  NORTHWEST
10           SCENIC ACRES                0.0  NORTHWEST
11             RANCHLANDS                0.0  NORTHWEST
12         SILVER SPRINGS                0.0  NORTHWEST
13            COLLINGWOOD                0.0  NORTHWEST
14              ROYAL OAK                0.0  NORTHWEST
15                    01F                0.0  NORTHWEST
16                VARSITY                0.0  NORTHWEST
17  UNIVERSITY OF CALGARY                0.0  NORTHWEST
18     UNIVERSITY HEIGHTS                0.0  NORTHWEST
19    UNIVERSITY DISTRICT                0.0  NORTHWEST
20             MONTGOMERY                0.0  NORTHWEST
21                TUSCANY                0.0  NORTHWEST
22            ROCKY RIDGE                0.0  NORTHWEST

In [73]:
northeast = clust1[clust1.SECTOR == 'NORTHEAST'].reset_index(drop=True)
northeast

Neighborhood  Indian Restaurant     SECTOR
0        MARLBOROUGH                0.0  NORTHEAST
1          HOMESTEAD                0.0  NORTHEAST
2           REDSTONE                0.0  NORTHEAST
3   MARLBOROUGH PARK                0.0  NORTHEAST
4        CORNERSTONE                0.0  NORTHEAST
5          PINERIDGE                0.0  NORTHEAST
6          ABBEYDALE                0.0  NORTHEAST
7          CITYSCAPE                0.0  NORTHEAST
8      CORAL SPRINGS                0.0  NORTHEAST
9         MARTINDALE                0.0  NORTHEAST
10   MAYLAND HEIGHTS                0.0  NORTHEAST
11      SADDLE RIDGE                0.0  NORTHEAST
12     SKYVIEW RANCH                0.0  NORTHEAST
13         WHITEHORN                0.0  NORTHEAST
14     MONTEREY PARK                0.0  NORTHEAST
15     VISTA HEIGHTS                0.0  NORTHEAST
16          TARADALE                0.0  NORTHEAST
17            TEMPLE                0.0  NORTHEAST